Validation and others for keeping variant override "On Demand" and "Count on Hand" compatible

kristinalim opened the pull request:
#### What? Why?

Relates to #3066 
Relates to #3067 
Relates to #3104 

This includes changes in #3104, which is a dependency for this PR.

The following changes are included:

1. Require VO "On Demand" and "Count on Hand" to be compatible. Only the following combinations are allowed for VOs now:
    * "Use producer stock settings", with no override value for "Count on Hand"
    * "Yes" on demand, with no override value for "Count on Hand"
    * "No" on demand, with required override value for "Count on Hand"
2. Migrate existing VOs to comply with these new validation rules.
3. Make resetting stock settings for VO change on demand to "No" (force limited stock).
4. Update Inventory data import to set appropriate VO "On Demand" for specified "On Hand" value.

For code review, please pay special attention to the DB migration, which would make the following changes based on the discussion in #3067 for existing VOs:

1. When "On Demand" is "Yes" but "Count on Hand" is not blank, clear "Count on Hand"
2. When "On Demand" is "Use producer stock settings" but "Count on Hand" is not blank, set "On Demand" to "No"
3. When "On Demand" is "No" but "Count on Hand" is blank, change "On Demand" to "Use producer stock settings"

#### What should we test?

Aside from testing the Inventory page UI changed in #3104, here are parts which should be tested:

1. Saving invalid VO should provide an appropriate error. For example, you may try setting "On Demand" to "No" but leaving "Count on Hand" blank.
2. Saving VO with the different allowed stock settings should be successful. Aside from checking the success notice, please reload the "Inventory" page and check the updated values.
3. Resetting stock settings for VOs with different allowed settings should be successful. This should set "On Demand" to "No" aside from updating to the default stock level.
4. Importing of inventory data with different combinations of stock settings, both valid and invalid. Check that the correct VO "Count on Hand" and "On Demand" settings are saved. 

#### Release notes

- Require variant override "On Demand" and "Count on Hand" to be compatible. Only the following combinations are allowed now:
    * On Demand: "Use producer stock settings", Count on Hand: blank
    * On Demand: "Yes", Count on Hand: blank
    * On Demand: "No", Count on Hand: filled
- Migrate existing variant overrides to comply with new validation rules. Changes to the variant overrides are recorded in `reports/SimplifyVariantOverrideStockSettings/` in the application path: `changed_variant_overrides.csv` has the complete list of affected records, and there are CSV files specific to each distributor affected.
- Make resetting stock settings for variant override change "On Demand" to "No" (`false`).
- Update Inventory data import to set appropriate variant override "On Demand" for specified "On Hand" value.

Changelog Category: Changed

#### Dependencies

This needs to be merged together with #3104.

mkllnk commented on app/models/product_import/entry_validator.rb:
```diff
@@ -323,8 +323,10 @@ def validate_inventory_item(entry, variant_override)
-      variant_override.assign_attributes(count_on_hand: entry.on_hand, import_date: @import_time)
-      check_on_hand_nil(entry, variant_override)
+
+      check_variant_override_stock_settings(entry, variant_override)
```
I find this very confusing.

1. count_on_hand and on_demand of the existing variant override are assigned based on entry.
2. entry.on_hand_nil is assigned based on the the variant override.
3. on_demand of the variant override is assigned again based on entry.

In this context, the entry is only used to provide `price` and `on_demand` for the variant_override. So what if `check_variant_override_stock_settings` was called `variant_override_attributes` and returned the right hash based on entry and existing variant override?

I guess that doesn't sanitise `entry` which could be used somewhere else. Am I right with my assumptions? Is there a way to avoid these circular assignments? I would find it easier if it was more linear, for example:

1. Derive truth from input variables, for example create hash from entry and override.
2. Assign values to variant override (which is concerned here).

Or if the entry needs to be modified:

1. Sanitise entry with help of variant override.
2. Assign values to variant override.

Sorry if this is confusing. I'm not sure if it's understandable what I'm trying to say. :-D.

Modifying state within objects is always more dangerous than having immutable data structures.

kristinalim replied:
I actually based this on the `check_on_hand_nil` method which it was originally using. I read through some of the import related code, but I confess I don't know how exactly `entry.on_hand_nil` is used. However, the attribute was already being set even before this PR: https://github.com/openfoodfoundation/openfoodnetwork/blob/df51d137dc87a819b3646e501798079ac387046e/app/models/product_import/entry_validator.rb#L354

`check_on_hand_nil` was setting presumed attributes for the VO if needed, and setting `entry.on_hand_nil` if there is no `on_hand` value.

@Matt-Yorkley Does this change seem safe enough to you?

If yes, @mkllnk are you okay with keeping that part as is for now?

mkllnk replied:
Yes, I'm okay with it.

kristinalim commented on db/migrate/20181128054803_simplify_variant_override_on_demand_and_count_on_hand.rb:
```diff
@@ -0,0 +1,29 @@
+    VariantOverride.where(on_demand: false, count_on_hand: nil).update_all(on_demand: nil)
+  end
+
+  def down; end
```
@mkllnk No, I haven't seen that yet. :+1: 

> Creating records for all affected rows would be good in case there are unknown side effects.

You are talking about creating a DB table for caching the original values, right? Yes, this would allow undoing the changes, which is safer.

Regarding naming of this table, I can shorten the DB migration to "SimplifyVariantOverrideStockSettings", and then name the table "simplify_variant_override_stock_settings_changed_variant_overrides" with `variant_override_id`, `on_demand`, and `count_on_hand` columns. Do let me know if this is what you had in mind.

Thank you!

sauloperez replied:
> You are talking about creating a DB table for caching the original values, right? Yes, this would allow undoing the changes, which is safer.

Isn't that a bit too much? I'm afraid something like this would cost us way too much on dev hours given the size of the change.

mkllnk replied:
No, I didn't mean creating another DB table. I meant creating a backup file with the affected records as in the example.

I do agree that it is a considerable overhead and added complexity. We shouldn't do that with every migration. I'm not sure about this one though. Yes, it is simple, but it potentially affects stock levels.

```rb
    VariantOverride.where(on_demand: nil).where("count_on_hand IS NOT NULL")
      .update_all(on_demand: false)
```

So if `on_demand` wasn't set, the variant's on_demand is considered. The stock logic may find that the variant is on demand and ignores the stock level of the override. So this would change the current logic, right?

```rb
VariantOverride.where(on_demand: false, count_on_hand: nil).update_all(on_demand: nil)
```

We are fixing invalid setups here, but they could be in use -- successfully. If the variant has `on_demand: false`, nothing changed. But if the variant is set `on_demand: true`, the stock level is not in use by the producer, but it is used by the overriding shop. They would lose their stock level. The shop may be sharing the stock level with another shop. I would doubt that they set it up intentionally like that. So we can probably copy the stock level from the variant to the override. But if they were sharing the stock level intentionally, this pull request would break their usage scenario.

We should query all production servers for these edge cases. Maybe we need to look at the affected enterprises and tell them about the change.

Changing stock levels is a big deal for enterprises. We had some drama with this recently: https://github.com/openfoodfoundation/openfoodnetwork/pull/2963
Following that incident I had the idea of backup files that allow the rollback of destructive migrations.

kristinalim replied:
Ah, I thought you were talking about providing an undo action for the DB migration.

Actually, based on my analysis in the issue description of #3067, the behaviour would not change, with the exception of this scenario:

* VO - OD: `false`, COH: `nil`
* Variant (producer) - OD: `true`, COH: filled

Here, instead of using the COH of the producer - which doesn't seem right because COH is not even visible in the product list, variant list, and variant form when the variant is "On Demand" - the DB migration sets it to use the producer's stock settings. (Noting that this quirk is the root of the problem in #3058.)

Aside from this scenario, do you see any other situation when the behaviour would change?

I agree that we should keep a backup file. So a CSV file with the data, so we can send to instance owners if needed.

I will put in the following data:

* Distributor Name
* Distributor ID
* Product Name
* Product ID
* Variant Description
* Variant ID
* Previous On Demand
* Previous Count on Hand
* Updated On Demand
* Updated Count on Hand

I'm already working on this, but let me know if you find issues.

kristinalim replied:
This is the script I used to generate the table in #3067:

```ruby
variant = Spree::Variant.new
variant_override = VariantOverride.new

def set_stock(object, on_demand, count_on_hand)
  object.on_demand = on_demand
  object.count_on_hand = count_on_hand
end

variant.extend(OpenFoodNetwork::ScopeVariantToHub::ScopeVariantToHub)
variant.instance_variable_set :@variant_override, variant_override

variant_configurations = [
  [true, 20],
  [false, 20]
]
variant_override_configurations = [
  [nil, nil],
  [true, nil],
  [false, nil],
  [nil, 10],
  [true, 10],
  [false, 10]
]

results = [["VO OD", "VO OH", "V OD", "V OH", "OD", "OH"]]

variant_configurations.each do |variant_conf|
  set_stock(variant, *variant_conf)

  variant_override_configurations.each do |variant_override_conf|
    set_stock(variant_override, *variant_override_conf)

    results << [
      *variant_override_conf,
      *variant_conf,
      variant.on_demand,
      (variant.count_on_hand unless variant.on_demand)
    ]
  end
end

results.sort_by(&:to_s).each_with_index do |result, i|
  puts "|#{([i] + result).map { |value| " #{value.inspect} |".rjust(8) }.join}"
end
```

kristinalim replied:
@mkllnk Let's continue discussing this in #3067.

kristinalim commented:
Kindly review this again.

1. Changes are now logged in `APP/reports/SimplifyVariantOverrideStockSettings-changed_variant_overrides.csv`.
2. The `down` method is now able to read the same report and undo the changes.

Here is a little script I wrote to test the migration: [20181214-test_simplify_variant_override_stock_settings.txt](https://github.com/openfoodfoundation/openfoodnetwork/files/2676906/20181214-test_simplify_variant_override_stock_settings.txt) (It uses the TXT extension so I can upload it in this comment.) You may review and load the script in your DEVELOPMENT Rails console. It should raise an error when the variant override values are not as expected after running `up` and after running `down` - I actually found a bug in the migration script with this.

kristinalim commented:
I have yet again updated the strategy for VOs with "On Demand" of `false` but blank "Count on Hand". Right now, we're setting these VOs to "Use producer stock settings". Please see [this comment onwards](https://github.com/openfoodfoundation/openfoodnetwork/issues/3067#issuecomment-446826259) for the reasoning.

@mkllnk I also added a README to `reports/`, and updated the `.gitignore` to ignore the directory except `/reports/README.md`.

This is the updated version of what I wrote to test the DB migration: 
[20181214-test_simplify_variant_override_stock_settings.txt](https://github.com/openfoodfoundation/openfoodnetwork/files/2678499/20181214-test_simplify_variant_override_stock_settings.txt)

kristinalim commented:
@mkllnk The shell script does not support commas in the enterprise name and include the CSV quotation marks. I made changes to the DB migration to also just generate reports by distributor.

I also updated the "Release Notes" above to reflect this.

sauloperez commented:
Staged on https://staging.katuma.org/

RachL commented:
@kristinalim I'm almost done testing but I need help on 2 things:
>Resetting stock settings for VOs with different allowed settings should be successful. This should set "On Demand" to "No" aside from updating to the default stock level.

The stock field is disabled, so it can't be done right?

About inventory import, I don't see any changes in the file template. Is that correct, can I just go and import the usual way?

myriamboure commented:
@RachL about inventory report the template is wrong apparently, and the import is not working, Sally found a bug. Check the PI project board if you want some details (sorry on the train hard to find the issue n°) So I guess you won't be able to test it and we'll test the bug fix from mat once this PR is merged so we can fix if there is an issue, maybe we would need to pay attention in testing inventory import... ping @Matt-Yorkley so you have that in mind.

kristinalim commented:
> > Resetting stock settings for VOs with different allowed settings should be successful. This should set "On Demand" to "No" aside from updating to the default stock level.
>
> The stock field is disabled, so it can't be done right?

@RachL Sorry, I'm realized what I said was confusing! I was talking about the "Reset Stock Levels to Defaults" feature, not "Use producer stock settings". Were you talking about that too and found a bug?

Actually, I only learned about this feature from working on this PR.

Here are more specific details for testing:

1. In the Inventory page, click "Enable Stock Reset?" under "Columns".
2. Enable stock reset for three variant overrides with different VO stock settings ("Use producer stock settings", "Yes", and "No"). Tick the checkbox under "Enable Stock Reset?", enter the default stock level, and then save.
3. Still in the Inventory page, under "Actions", click on "Reset Stock Levels to Defaults".
4. Check that the VOs now have "No" for "On Demand", and that their new count on hand are the stock levels set in Step 2.

luisramos0 commented:
Moved this one from ES to FR staging so I can use ES for another PR (confirmed with @RachL on slack).

RachL commented:
@kristinalim do you agree that there is a step missing between 2 and 3 ? 

1. In the Inventory page, click "Enable Stock Reset?" under "Columns".
2. Enable stock reset for three variant overrides with different VO stock settings ("Use producer stock settings", "Yes", and "No"). Tick the checkbox under "Enable Stock Reset?", enter the default stock level, and then save.
**3. Change the stock settings of these 3 variant overrides**
4. Still in the Inventory page, under "Actions", click on "Reset Stock Levels to Defaults".
5. Check that the VOs now have "No" for "On Demand", and that their new count on hand are the stock levels set in Step 2.

Otherwise, when resetting on step 4, we keep the current settings as we just set up, right? However I still can't see _Check that the VOs now have "No" for "On Demand"_ indeed if something was set to on demand.... it stays "on demand".... :( Am I missing something?

kristinalim commented:
I don't think this is necessary: "Change the stock settings of these 3 variant overrides" I'm not sure - maybe there is a bug? Although it worked when I did this on my development server. I will make a video recording of how it works for me and post that here.

Thanks, @RachL! :slightly_smiling_face:

kristinalim commented:
@RachL Here it is. In this recording, I was using "Beef", "Carrots", and "Potatoes" 1g :grin: variant (not the 1kg variant). Note that initially they had different stock settings. After resetting stock levels to default, they all had "No" in "On Demand" and their COH have been changed.

![20181219230100-reset-stock-levels-to-default](https://user-images.githubusercontent.com/2243/50228050-1c12d800-03e2-11e9-8584-72ed23ee9adb.gif)

kristinalim commented:
Ah, I think I understand your concern, @RachL! Clicking the "Reset Stock Levels to Defaults" action already saves the data in the server - you don't need to save the changes in the fields again through the "Save Changes" button.

RachL commented:
aaaaaah yes thank you so much @kristinalim this all good, we can move it to ready to go :tada: and keeping in my mind that we will need to be carefull with inventory import once it is solved. thanks @myriamboure for the heads up

https://docs.google.com/document/d/1qOENjrcc9PWFvy7NoFgR42ecj91BaeYgOCZcV1xa2V8/edit#
